test/integration/BLACKBOX_README.md

```diff
@@ -0,0 +1,398 @@
+# TEAL Blackbox Toolkit: Program Reporting and Testing via Dry Runs
+
+**NOTE: to get math formulas to render here using Chrome, add the [xhub extension](https://chrome.google.com/webstore/detail/xhub/anidddebgkllnnnnjfkmjcaallemhjee/related) and reload**
+
+## Blackbox Testing Howto
+
+### What is TEAL Blackbox Testing?
+
+TEAL Blackbox Testing lets you treat your TEAL program as a black box that for every received input produces an output and other observable effects. You can create reports that summarize those effects, and turn the _reports_ into _program invariant conjectures_ which you then check with _sequence assertions_.
+
+### Why Blackbox Testing?
+
+Here are some use cases:
+
+* by allowing you to assert that certain invariants hold over a large set of inputs you gain greater confidence that your TEAL programs and AVM smart contracts work as designed
+* when tweaking, refactoring or optimizing your TEAL source, ensure that no regressions have occured
+* allows AVM developers to practice the art of TTDD (TEAL Test Driven Development)
+
+## Simple TEAL Blackbox Toolkit Example: Program for $`x^2`$
+
+Consider the following [TEAL program](https://github.com/algorand/py-algorand-sdk/blob/23c21170cfb19652d5da854e499dca47eabb20e8/x/blackbox/teal/lsig_square.teal) that purportedly computes $`x^2`$:
+
+```plain
+#pragma version 6
+arg 0
+btoi
+callsub square_0
+return
+
+// square
+square_0:
+store 0
+load 0
+pushint 2 // 2
+exp
+retsub
+```
+
+You'd like to write some unit tests to validate that it computes what you think it should, and also make **assertions** about the:
+
+* program's opcode cost
+* program's stack
+* stack's height
+* scratch variables
+* final log message (this is especially useful for [ABI-compliant programs](https://developer.algorand.org/docs/get-details/dapps/smart-contracts/ABI/))
+* status (**PASS**, **REJECT** or _erroring_)
+* error conditions that are and aren't encountered
+
+Even better, before making fine-grained assertions you'd like to get a sense of what the program is doing on a large set of inputs so you can discover program invariants. The TEAL Blackbox Toolkit's recommended approach for enabling these goals is to:
+
+* start by making basic assertions and validate them using dry runs (see "**Basic Assertions**" section below)
+* execute the program on a run-sequence of inputs and explore the results (see "**EDRA: Exploratory Dry Run Analysis**" section below)
```

Comment:
what does "run-sequence" mean? Is it the same as a "sequence" of inputs or does it have additional meaning?

Reply:
yeah, just a sequence of inputs.